### PR TITLE
fix(fuzzing): persist snapshot failures when fuzzing

### LINKED ISSUE #3055: bug: snapshot/revertTo causes tests to pass unexpectedly
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (48d5d79 2022-09-01T00:10:22.933552Z)

### What command(s) is the bug in?

forge test

### Operating System

_No response_

### Describe the bug

If you run this test, it passes. If you comment out the `vm.revertTo` line it fails. Failure is expected in both cases since the assertion should fail. It seems like the cause is that reverting the state also reverts the `failed` flag: https://github.com/dapphub/ds-test/blob/9310e879db8ba3ea6d5c6489a579118fd264a3f5/src/test.sol#L47-L63
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.15;
import "forge-std/Test.sol";

contract MyTest is Test {
  function test_snapshot() external {
    uint256 snapId = vm.snapshot();
    assertEq(uint256(0), uint256(1));
    vm.revertTo(snapId);
  }
}
```

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

Closes #3055.

Fuzzing was having soundness issues with snapshot failures. This is because snapshot failures were not being persisted throughout fuzz runs, and when the Fuzz Backend Wrapper we use was dropped, this information was lost, leading to inconsistent test PASS states.

## Solution

Persist snapshot failures when fuzzing.

### CODE REVIEW

File: evm/src/executor/backend/mod.rs [hunk 1]
```diff
@@ -848,7 +858,7 @@ impl DatabaseExt for Backend {
                 .map(|addr| self.is_failed_test_contract_state(addr, current_state))
                 .unwrap_or_default()
             {
-                self.inner.has_snapshot_failure = true;
+                self.inner.has_snapshot_failure.fetch_or(true, Ordering::Relaxed);
```
Comment (mattsse) [inline evm/src/executor/backend/mod.rs:861]: this is just store, right?
```suggestion
                self.inner.has_snapshot_failure.store(true, Ordering::SeqCst);
```

**Review by mattsse — CHANGES_REQUESTED** [review]
incredible find!

using the atomic to detect failures across fuzz runs is great!

smol nit

File: evm/src/executor/backend/mod.rs (hunk 1 above)
Comment (Evalir) [reply evm/src/executor/backend/mod.rs:861]: oh yeah should be, good catch

Comment (Evalir) [reply evm/src/executor/backend/mod.rs]: done in https://github.com/foundry-rs/foundry/pull/4974/commits/7fb915a36687bc54e9a32b4637fd2b194762d432

**Pull request merged**